### PR TITLE
Add support for delta timesteps streaming

### PR DESCRIPTION
**Pull request by fracapuano**
Part of PR stack
👉 https://github.com/huggingface/lerobot/pull/1173
👀 https://github.com/huggingface/lerobot/pull/1165
👀 https://github.com/huggingface/lerobot/pull/969

## What this does

This PR extends `StreamingLeRobotDataset` to support delta timestamps, enabling efficient access to past and future frames during streaming without loading the entire dataset into memory.

### Enhanced Backtrackable Iterator
- **Bidirectional buffering**: Maintains separate buffers for history (`_back_buf`) and lookahead (`_ahead_buf`)
- **Configurable windows**: Set `history` and `lookahead` parameters to control buffer sizes. These parameters are inferred automatically from the input `delta_timesteps` according to a worst-case-scenario logic (min and max delta values, respectively).
- **Memory efficient**: Only keeps necessary frames in memory based on delta timestamp requirements
- **Episode-aware**: Prevents accessing frames across episode boundaries, returning padding and padding masks when within-same-episode values are not accessible.

### Delta Timestamps Integration
- **Automatic validation**: Checks delta timestamp keys against dataset features at initialization
- **Temporal queries**: Supports both past (`delta < 0`) and future (`delta > 0`) frame access
- **Padding handling**: Automatically pads with zeros when frames are unavailable or cross episode boundaries. Returns the padding masks as a `torch.Bool` mask within `{key}.pad_masking` for easy post-filtering of the returned elements
- **Video compatibility**: Works seamlessly with video decoding for visual modalities. Relies on `torchcodec` multi-timesteps integration.

### Backtrackable Class (`utils.py`)
```python
# Create iterator with 5-frame history and 3-frame lookahead
backtrackable = Backtrackable(dataset, history=5, lookahead=3)

# Access patterns
current = next(backtrackable)           # Move forward
previous = backtrackable.peek_back(1)   # Look back without moving
future = backtrackable.peek_ahead(2)    # Look ahead without moving
```

**Core methods**:
- `peek_back(n)`: Access n frames back without cursor movement
- `peek_ahead(n)`: Access n frames ahead, pre-fetching if needed
- `can_peek_back/ahead()`: Check availability before access
- Episode boundary protection via custom exceptions

### Streaming Dataset Extensions
- **Window calculation**: Automatically determines required buffer sizes from delta timestamps
- **Frame stacking**: Assembles multi-temporal observations into single items
- **Graceful degradation**: Falls back to zero-padding when frames unavailable
- **Performance optimization**: Minimal memory footprint with bounded buffers

## How it was tested
Tested using the `profile_streaming_dataset.py` profiling script (complete report below 👇)

## How to checkout & try? (for the reviewer)
Checkout to the branch, install the correct dependancies (pay particular attention to `torchcodec`, cfr. #1165) and then execute this: 
```python
delta_timestamps = {
    "observation.images.cam_right_wrist": [-0.1, -0.05, 0.0],  # 100ms, 50ms ago, current
    "action": [0.0, 0.02, 0.04],  # current, 20ms, 40ms future
}

dataset = StreamingLeRobotDataset(
    repo_id="lerobot/aloha_mobile_cabinet",
    delta_timestamps=delta_timestamps
)

for item in dataset:
    # item contains stacked frames according to delta_timestamps, with padding information for easier filtering
    print(item["action"].shape)  # (3, action_dim), as we're quering for 3 elements. Same for `observation`
    print(item["action.pad_masking"])  # torch.tensor([False, False, True]), in this example first two elements are *not* padded, third one is
```

cc. @lhoestq for review and @Cadene for viz

### CODE REVIEW

**Comment by fracapuano** [comment]
# Assessing performance

Performance metrics for this updated version of `StreamingLeRobotDataset` are (1) throughput and (2) sample randomness (cfr. #1165).

## Profiling results ⏱️
```bash
TIMING STATISTICS
============================================================
Warmup timing (ms):
  Mean: 15057.48 ± 0.00
  Median: 15057.48
  Range: [15057.48, 15057.48]

Per-sample timing (ms):
  Mean: 21.91 ± 313.48
  Median: 13.69
  Range: [3.26, 15051.73]

Per-run timing (seconds):
  Mean: 60.11 ± 0.91
  Range: [58.80, 61.28]

Throughput:
  Samples/second: 33.27
============================================================

DETAILED LINE PROFILING RESULTS
============================================================
Timer unit: 1e-09 s

Total time: 42.6932 s
File: /Users/francescocapuano/Desktop/lerobot/lerobot/common/datasets/profile_streaming_dataset.py
Function: iterate_dataset at line 171

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   171                                               def iterate_dataset(ds, n):
   172                                                   # Iterating without warmup for line profiling
   173         1       2000.0   2000.0      0.0          iterator = iter(ds)
   174         1       1000.0   1000.0      0.0          start_time = time.time()
   175      2001     547000.0    273.4      0.0          for _ in range(n):
   176      2000        4e+10    2e+07    100.0              next(iterator)
   177         1       1000.0   1000.0      0.0          end_time = time.time()
   178         1       1000.0   1000.0      0.0          return end_time - start_time

Total time: 42.6847 s
File: /Users/francescocapuano/Desktop/lerobot/lerobot/common/datasets/streaming_dataset.py
Function: __iter__ at line 189

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   189                                               def __iter__(self) -> Iterator[Dict[str, torch.Tensor]]:
   190         1       1000.0   1000.0      0.0          buffer_indices_generator = self._iter_random_indices(self.rng, self.buffer_size)
   191         2    1364000.0 682000.0      0.0          idx_to_backtracktable_dataset = {
   192                                                       idx: self._make_backtrackable_dataset(self.hf_dataset.shard(self.num_shards, index=idx))
   193         1          0.0      0.0      0.0              for idx in range(self.num_shards)
   194                                                   }
   195                                           
   196                                                   # This buffer is populated while iterating on the dataset's shards
   197         1          0.0      0.0      0.0          frames_buffer = []
   198         1          0.0      0.0      0.0          try:
   199      3000    4442000.0   1480.7      0.0              while available_shards := list(idx_to_backtracktable_dataset.keys()):
   200      3000   19098000.0   6366.0      0.0                  shard_key = next(self._infinite_generator_over_elements(available_shards))
   201      3000     633000.0    211.0      0.0                  dataset = idx_to_backtracktable_dataset[shard_key]  # selects which shard to iterate on
   202      3000        4e+10    1e+07     97.0                  for frame in self.make_frame(dataset):
   203      3000    1526000.0    508.7      0.0                      if len(frames_buffer) == self.buffer_size:
   204      2000    4945000.0   2472.5      0.0                          i = next(buffer_indices_generator)
   205      2000     869000.0    434.5      0.0                          yield frames_buffer[i]
   206      1999 1234243000.0 617430.2      2.9                          frames_buffer[i] = frame
   207                                                               else:
   208      1000     421000.0    421.0      0.0                          frames_buffer.append(frame)
   209      2999    6899000.0   2300.4      0.0                      break  # random shard sampled, switch shard
   210                                           
   211         2          0.0      0.0      0.0          except (
   212         1          0.0      0.0      0.0              RuntimeError,
   213         1          0.0      0.0      0.0              StopIteration,
   214                                                   ):  # NOTE: StopIteration inside a generator throws a RuntimeError since 3.7
   215                                                       del idx_to_backtracktable_dataset[shard_key]  # Remove exhausted shard, onto another shard
   216                                           
   217                                                   # Once shards are all exhausted, shuffle the buffer and yield the remaining frames
   218                                                   self.rng.shuffle(frames_buffer)
   219                                                   yield from frames_buffer

Total time: 7.1e-05 s
File: /Users/francescocapuano/Desktop/lerobot/lerobot/common/datasets/streaming_dataset.py
Function: _make_backtrackable_dataset at line 221

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   221                                               def _make_backtrackable_dataset(self, dataset: datasets.IterableDataset) -> Backtrackable:
   222         1      60000.0  60000.0     84.5          history, lookahead = self._get_window_steps()
   223         1      11000.0  11000.0     15.5          return Backtrackable(dataset, history=history, lookahead=lookahead)

Total time: 41.3923 s
File: /Users/francescocapuano/Desktop/lerobot/lerobot/common/datasets/streaming_dataset.py
Function: make_frame at line 225

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   225                                               @profile
   226                                               def make_frame(self, dataset_iterator: Backtrackable) -> Generator:
   227                                                   """Makes a frame starting from a dataset iterator"""
   228      3000 1769371000.0 589790.3      4.3          item = next(dataset_iterator)
   229      3000   15610000.0   5203.3      0.0          item = item_to_torch(item)
   230                                           
   231                                                   # Get episode index from the item
   232      3000     679000.0    226.3      0.0          ep_idx = item["episode_index"]
   233                                           
   234                                                   # Apply delta querying logic if necessary
   235      3000     460000.0    153.3      0.0          if self.delta_indices is not None:
   236      3000  835693000.0 278564.3      2.0              query_result, padding = self._get_delta_frames(dataset_iterator, item)
   237      3000    2359000.0    786.3      0.0              item = {**item, **query_result, **padding}
   238                                           
   239                                                   # Load video frames, when needed
   240      3000    6654000.0   2218.0      0.0          if len(self.meta.video_keys) > 0:
   241      3000     656000.0    218.7      0.0              current_ts = item["timestamp"]
   242      3000     535000.0    178.3      0.0              query_indices = self.delta_indices if self.delta_indices is not None else None
   243      3000   68575000.0  22858.3      0.2              query_timestamps = self._get_query_timestamps(current_ts, query_indices)
   244      3000        4e+10    1e+07     93.0              video_frames = self._query_videos(query_timestamps, ep_idx)
   245      3000    6970000.0   2323.3      0.0              item = {**video_frames, **item}
   246                                           
   247      3000  206320000.0  68773.3      0.5          item["task"] = self.meta.tasks.iloc[item["task_index"]].name
   248                                           
   249      3000     619000.0    206.3      0.0          yield item


RANDOMNESS ANALYSIS
============================================================
Correlation between iteration index and frame index: 0.0772
(Correlation close to 0 indicates more random access pattern)
(Correlation close to 1 indicates sequential access pattern)
============================================================

``` 

- 📝 Peak throughput averages around 80fps, with mean throughput almost halved versus non-delta timesteps. I suspect this can be mitigated implementing some optimizations on the frame retrieval logic (**not** relatively to `torchcodec`)
- 📝 As the randomness mechanism is left unchanged, randomness is left ~untouched versus #1165

**Comment by fracapuano** [comment]
Hey @lhoestq 👋 I think appropriate values for history/lookahead are to play this safe is somewhere 50-100 (probably better to do 100 and play it on the safe side)

File: .cache/calibration/so100/main_follower.json
Comment (AdilZouitine) [inline .cache/calibration/so100/main_follower.json:1]: This file should be removed

File: .cache/calibration/so100/main_leader.json
Comment (AdilZouitine) [inline .cache/calibration/so100/main_leader.json:1]: This file should be removed
